{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The script:\n",
    " - selects non-overlapping & single CDS genes,\n",
    " - extracts CDSs of non-overlapping & single CDS genes (this is done with seqtk & bedtools),\n",
    " - reads fasta files with sequences of genes extracted from the genome fasta files,\n",
    " - reverse complements genes,\n",
    " - creates multiple alignments with concatenated genes.\n",
    " #### @author:aniafijarczyk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import glob\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import generic_dna\n",
    "import pandas as pd\n",
    "from collections import defaultdict\n",
    "from urllib.request import urlopen\n",
    "import gzip\n",
    "import re\n",
    "from Bio.Seq import Seq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Selecting non-overlapping genes\n",
    "Downloading gff file from SGD database (this is version R64-2-1); \n",
    "reading gff file; extracting information for genes, including modified chromosome name, start and stop in bed format and gene ID. Excluding genes which have non-overlapping gene coordinates (it is possible that some genes may be overlapping but their CDS are not, but I'll ignore it here)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>geneID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ref|NC_001133|</td>\n",
       "      <td>334</td>\n",
       "      <td>649</td>\n",
       "      <td>YAL069W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ref|NC_001133|</td>\n",
       "      <td>537</td>\n",
       "      <td>792</td>\n",
       "      <td>YAL068W-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ref|NC_001133|</td>\n",
       "      <td>1806</td>\n",
       "      <td>2169</td>\n",
       "      <td>YAL068C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ref|NC_001133|</td>\n",
       "      <td>2479</td>\n",
       "      <td>2707</td>\n",
       "      <td>YAL067W-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ref|NC_001133|</td>\n",
       "      <td>7234</td>\n",
       "      <td>9016</td>\n",
       "      <td>YAL067C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            chrom  start  stop     geneID\n",
       "0  ref|NC_001133|    334   649    YAL069W\n",
       "1  ref|NC_001133|    537   792  YAL068W-A\n",
       "2  ref|NC_001133|   1806  2169    YAL068C\n",
       "3  ref|NC_001133|   2479  2707  YAL067W-A\n",
       "4  ref|NC_001133|   7234  9016    YAL067C"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# this should be version R64-2-1\n",
    "url = \"http://sgd-archive.yeastgenome.org/curation/chromosomal_feature/saccharomyces_cerevisiae.gff.gz\"\n",
    "fh = gzip.open(urlopen(url),'rt')\n",
    "linie = fh.readlines()\n",
    "F = []\n",
    "for ele in linie:\n",
    "    if ele != \"##FASTA\\n\":\n",
    "        if ele.startswith(\"#\"):\n",
    "            continue\n",
    "        else:\n",
    "            F.append(ele.split())\n",
    "    else:\n",
    "        break\n",
    "dF = list(zip(*F))\n",
    "df = pd.DataFrame({\"chrom\":dF[0],\"source\":dF[1],\"feature\":dF[2],\"start\":dF[3],\"end\":dF[4],\"score\":dF[5],\n",
    "                  \"strand\":dF[6],\"frame\":dF[7],\"attribute\":dF[8]})\n",
    "df_genes = df.loc[df['feature'] == \"gene\",[\"chrom\",\"start\",\"end\",\"attribute\"]]\n",
    "df_genes[\"geneID\"] = df_genes[\"attribute\"].apply(lambda x: re.split(\"[=,;]\",x)[1])\n",
    "df_genes['newstart'] = df_genes['start'].apply(lambda x: int(x)-1)\n",
    "df_genes['newend'] = df_genes['end'].apply(lambda x: int(x))\n",
    "ref_names = {\"chrI\":\"ref|NC_001133|\",\"chrII\":\"ref|NC_001134|\",\n",
    "\"chrIII\":\"ref|NC_001135|\",\"chrIV\":\"ref|NC_001136|\",\n",
    "\"chrV\":\"ref|NC_001137|\",\"chrVI\":\"ref|NC_001138|\",\n",
    "\"chrVII\":\"ref|NC_001139|\",\"chrVIII\":\"ref|NC_001140|\",\n",
    "\"chrIX\":\"ref|NC_001141|\",\"chrX\":\"ref|NC_001142|\",\n",
    "\"chrXI\":\"ref|NC_001143|\",\"chrXII\":\"ref|NC_001144|\",\n",
    "\"chrXIII\":\"ref|NC_001145|\",\"chrXIV\":\"ref|NC_001146|\",\n",
    "\"chrXV\":\"ref|NC_001147|\",\"chrXVI\":\"ref|NC_001148|\",\"chrmt\":\"ref|NC_001224|\"}\n",
    "df_genes[\"newchrom\"] = df_genes[\"chrom\"].apply(lambda x: ref_names[x])\n",
    "bed = df_genes[['newchrom','newstart','newend','geneID']].rename(columns={\"newchrom\":\"chrom\",\"newstart\":\"start\",\"newend\":\"stop\"}).reset_index().drop(columns=[\"index\"])\n",
    "bed.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### List of non-overlapping genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "chroms = sorted(list(set(bed['chrom'])))\n",
    "Ch = {}\n",
    "C = {}\n",
    "Genes_nono = []\n",
    "for chrom in chroms:\n",
    "    db = bed.loc[bed[\"chrom\"]==chrom,:].reset_index()\n",
    "    starts = list(db[\"start\"])\n",
    "    stops = list(db[\"stop\"])\n",
    "    pairs = zip(starts,stops)\n",
    "    ranges = [set(range(int(i[0]),int(i[1]))) for i in pairs]\n",
    "    Ch[chrom] = ranges\n",
    "    first_start = int(db.at[0,'start'])\n",
    "    first_stop = int(db.at[0,'stop'])\n",
    "    geneID = db.at[0,'geneID']\n",
    "    #firstset = set(range(int(first_start),int(first_stop)))\n",
    "    F = [(first_start, first_stop, geneID)]\n",
    "    for index, row in db.iterrows():\n",
    "        i1 = F[-1][0]\n",
    "        i2 = F[-1][1]\n",
    "        lastset = set(range(i1,i2))\n",
    "        newcoords = int(row['start']), int(row['stop']), row['geneID']\n",
    "        newrange = set(range(int(row['start']), int(row['stop'])))\n",
    "        if len(lastset.intersection(newrange))>0:\n",
    "            continue\n",
    "        else:\n",
    "            F.append(newcoords)\n",
    "            Genes_nono.append(row['geneID'])\n",
    "    C[chrom] = F"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Selecting cds of non-overlapping & single CDS genes and writing to bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>cdsID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ref|NC_001133|</td>\n",
       "      <td>1806</td>\n",
       "      <td>2169</td>\n",
       "      <td>YAL068C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ref|NC_001133|</td>\n",
       "      <td>2479</td>\n",
       "      <td>2707</td>\n",
       "      <td>YAL067W-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ref|NC_001133|</td>\n",
       "      <td>7234</td>\n",
       "      <td>9016</td>\n",
       "      <td>YAL067C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ref|NC_001133|</td>\n",
       "      <td>10090</td>\n",
       "      <td>10399</td>\n",
       "      <td>YAL066W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ref|NC_001133|</td>\n",
       "      <td>11564</td>\n",
       "      <td>11951</td>\n",
       "      <td>YAL065C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            chrom  start   stop      cdsID\n",
       "0  ref|NC_001133|   1806   2169    YAL068C\n",
       "1  ref|NC_001133|   2479   2707  YAL067W-A\n",
       "2  ref|NC_001133|   7234   9016    YAL067C\n",
       "3  ref|NC_001133|  10090  10399    YAL066W\n",
       "4  ref|NC_001133|  11564  11951    YAL065C"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_cds = df.loc[df['feature'] == \"CDS\",[\"chrom\",\"start\",\"end\",\"attribute\",\"strand\"]]\n",
    "df_cds[\"cdsID\"] = df_cds[\"attribute\"].apply(lambda x: re.split(\"[=,;,_]\",x)[1])\n",
    "df_cds['newstart'] = df_cds['start'].apply(lambda x: int(x)-1)\n",
    "df_cds['newend'] = df_cds['end'].apply(lambda x: int(x))\n",
    "df_cds[\"newchrom\"] = df_cds[\"chrom\"].apply(lambda x: ref_names[x])\n",
    "df_cds[\"name\"] = df_cds.apply(lambda x: str(x[\"cdsID\"])+\"_\"+str(x[\"newchrom\"])+\"_\"+str(x[\"newstart\"])+\"_\"+str(x[\"newend\"]), axis=1)\n",
    "df_cdsf = df_cds.loc[df_cds['cdsID'].isin(Genes_nono),:]\n",
    "\n",
    "dups = []\n",
    "alls = []\n",
    "for ele in list(df_cdsf['cdsID']):\n",
    "    if ele in alls:\n",
    "        dups.append(ele)\n",
    "    else:\n",
    "        alls.append(ele)\n",
    "sing = [ele for ele in alls if ele not in list(set(dups))]\n",
    "df_cdsff = df_cdsf.loc[df_cdsf['cdsID'].isin(sing),:]\n",
    "bed_cds = df_cdsff[['newchrom','newstart','newend','cdsID']].rename(columns={\"newchrom\":\"chrom\",\"newstart\":\"start\",\"newend\":\"stop\"}).reset_index().drop(columns=[\"index\"])\n",
    "bed_cds.to_csv(\"./output/manipulateFasta_nonoverlappingCDS.bed\",sep=\"\\t\",header=False, index=False)\n",
    "bed_cds.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Fasta files with CDS extracted from strain genomes\n",
    "#### Getting table with strain genotypes from vcf (relatives_annot_Filtered2.vcf.gz = File S5)\n",
    "```console\n",
    "bcftools query -f '%CHROM\\t%POS[\\t%TGT]\\n' relatives_annot_Filtered2.vcf.gz | sed 's/\\//\\|/g' \\\n",
    "| awk  -F\"\\t\" -v OFS=\"\\t\" 'function GSUB(F) {gsub(/[|]/,\"\\t\",$F)} {GSUB(3);GSUB(4);GSUB(5);GSUB(6);GSUB(7);GSUB(8);GSUB(9);GSUB(10);GSUB(11);GSUB(12);GSUB(13)}1' \\\n",
    "| awk '{if (length($3)==1) print $0}' > relatives_annot_Filtered2.tab\n",
    "sed -i -e 's/\\./N/g' relatives_annot_Filtered2.tab\n",
    "```\n",
    "\n",
    "#### Generating fasta sequence for each strain genome & extracting CDSs from each strain genome\n",
    "```console\n",
    "i=0\n",
    "while read h\n",
    "  do\n",
    "  (( i = i + 1 ))\n",
    "  echo $i\n",
    "  awk -v var=\"$i\" '{print $1\"\\t\"$2\"\\t.\\t\"$(var+2)}' relatives_annot_Filtered2.tab > genotypes_temp.tab\n",
    "  seqtk mutfa S288C_reference_sequence_R64-2-1_20150113_N.fasta genotypes_temp.tab > refScer_${h}.fasta\n",
    "  bedtools getfasta -fi refScer_${h}.fasta -bed manipulateFasta_nonoverlappingCDS.bed -name -fo cds_${h}.fasta\n",
    "  done<relatives_annot_Filtered2.samples\n",
    "```\n",
    "##### relatives_annot_Filtered2.samples - names of strain genomes in relatives_annot_Filtered2.tab file\n",
    "##### S288C_reference_sequence_R64-2-1_20150113_N.fasta - reference chromosomes masked with N"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reverse complementing & concatenating CDS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./cds/cds_A.2565_H1.fasta 8010768\n",
      "./cds/cds_A.2565_H2.fasta 8010768\n",
      "./cds/cds_A.2565_H3.fasta 8010768\n",
      "./cds/cds_A.2565_H4.fasta 8010768\n",
      "./cds/cds_A.Muntons_H1.fasta 8010768\n",
      "./cds/cds_A.Muntons_H2.fasta 8010768\n",
      "./cds/cds_A.Muntons_H3.fasta 8010768\n",
      "./cds/cds_A.Muntons_H4.fasta 8010768\n",
      "./cds/cds_A.S-33_H1.fasta 8010768\n",
      "./cds/cds_A.S-33_H2.fasta 8010768\n",
      "./cds/cds_A.S-33_H3.fasta 8010768\n",
      "./cds/cds_A.S-33_H4.fasta 8010768\n",
      "./cds/cds_A.T-58_H1.fasta 8010768\n",
      "./cds/cds_A.T-58_H2.fasta 8010768\n",
      "./cds/cds_A.T-58_H3.fasta 8010768\n",
      "./cds/cds_A.T-58_H4.fasta 8010768\n",
      "./cds/cds_A.Windson_H1.fasta 8010768\n",
      "./cds/cds_A.Windson_H2.fasta 8010768\n",
      "./cds/cds_A.Windson_H3.fasta 8010768\n",
      "./cds/cds_A.Windson_H4.fasta 8010768\n",
      "./cds/cds_BE005_H1.fasta 8010768\n",
      "./cds/cds_BE005_H2.fasta 8010768\n",
      "./cds/cds_BE005_H3.fasta 8010768\n",
      "./cds/cds_BE005_H4.fasta 8010768\n",
      "./cds/cds_CFI_H1.fasta 8010768\n",
      "./cds/cds_CFI_H2.fasta 8010768\n",
      "./cds/cds_CFI_H3.fasta 8010768\n",
      "./cds/cds_CFI_H4.fasta 8010768\n",
      "./cds/cds_CFN_H1.fasta 8010768\n",
      "./cds/cds_CFN_H2.fasta 8010768\n",
      "./cds/cds_CFN_H3.fasta 8010768\n",
      "./cds/cds_CFN_H4.fasta 8010768\n",
      "./cds/cds_CFP_H1.fasta 8010768\n",
      "./cds/cds_CFP_H2.fasta 8010768\n",
      "./cds/cds_CFP_H3.fasta 8010768\n",
      "./cds/cds_CFP_H4.fasta 8010768\n",
      "./cds/cds_CHK_H1.fasta 8010768\n",
      "./cds/cds_CHK_H2.fasta 8010768\n",
      "./cds/cds_Jean-Talon_H1.fasta 8010768\n",
      "./cds/cds_Jean-Talon_H2.fasta 8010768\n",
      "./cds/cds_Jean-Talon_H3.fasta 8010768\n",
      "./cds/cds_Jean-Talon_H4.fasta 8010768\n"
     ]
    }
   ],
   "source": [
    "# File with list of sample names from relatives_annot_Filtered2.vcf (FileS5.vcf) file\n",
    "samp_file = pd.read_csv(\"./input_files/relatives_annot_Filtered2.samples\", sep = \"\\t\",header=None,names=[\"haplo\"])\n",
    "samples = list(samp_file[\"haplo\"])\n",
    "group = \"all\"\n",
    "\n",
    "###In case of selecting a subset of samples\n",
    "samp_to_filter = [\"A.2565\",\"A.Muntons\",\"A.S-33\",\"A.T-58\",\"BE005\",\"CFI\",\"CFN\",\"CFP\",\"CHK\",\"Jean-Talon\"]\n",
    "samp_to_filter = [\"Jean-Talon\"]\n",
    "group = samp_to_filter[0]\n",
    "samples_to_filter = [ele for ele in samples if ele.split(\"_\")[0] in samp_to_filter]\n",
    "#samples = samples_to_filter\n",
    "\n",
    "wh = open(\"genesConcatenated_pair_\"+group+\".fasta\",\"w\")\n",
    "for haplotype in samples:\n",
    "    fasta = glob.glob(\"./cds/cds_\"+haplotype+\".fasta\")[0]\n",
    "    records = [rec for rec in list(SeqIO.parse(fasta,\"fasta\"))]\n",
    "    R={}\n",
    "    CON = []\n",
    "    for gene in records:\n",
    "        if len(str(gene.seq))/3. == int(len(str(gene.seq))/3.):\n",
    "            if gene.id.split(\"-\")[0][-1] == 'C':\n",
    "                revcomp = gene.seq.reverse_complement()\n",
    "                R[gene.id] = str(revcomp)\n",
    "                CON.append(str(revcomp))\n",
    "            else:\n",
    "                R[gene.id] = str(gene.seq)\n",
    "                CON.append(str(gene.seq))\n",
    "        else:\n",
    "            continue\n",
    "    concatenated = ''\n",
    "    for s in CON:\n",
    "        concatenated += s\n",
    "\n",
    "    if haplotype.split(\"_\")[0]!=\"A.Windson\":\n",
    "        wh.write(\">\"+haplotype+\"\\n\")\n",
    "        wh.write(concatenated+\"\\n\")\n",
    "    print(fasta, len(concatenated))\n",
    "    \n",
    "wh.flush()\n",
    "wh.close()  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5713"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(bed_cds)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "base"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
